test/jest/Bladeburner/Actions.test.ts

```diff
@@ -0,0 +1,358 @@
+    player.startBladeburner();
+    return true;
+  }
+
+  function basicStats() {
+    inst.rank = 1;
+    inst.changeRank(Player, 10);
+    Player.gainStrengthExp(LOW_STAT_EXP);
+    Player.gainDefenseExp(LOW_STAT_EXP);
+    Player.gainAgilityExp(LOW_STAT_EXP);
+    Player.gainDexterityExp(LOW_STAT_EXP);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
+
+  function guaranteeSuccess() {
+    inst.rank = 1;
+    inst.changeRank(Player, 1e10);
+
+    Player.gainStrengthExp(HIGH_STAT_EXP);
+    Player.gainDefenseExp(HIGH_STAT_EXP);
+    Player.gainAgilityExp(HIGH_STAT_EXP);
+    Player.gainDexterityExp(HIGH_STAT_EXP);
+    inst.setSkillLevel(BladeburnerSkillName.BladesIntuition, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.Cloak, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.DigitalObserver, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.Reaper, 1e12);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
+
+  function guaranteeFailure() {
+    inst.rank = 1;
+    inst.changeRank(Player, 400e3); // Minimum to attempt to fail hardest op
+    Player.gainAgilityExp(1e50);
+    inst.calculateMaxStamina();
+    inst.stamina = 0;
+
+    resetCity();
+  }
+
+  function resetCity() {
+    inst.cities[inst.city].chaos = 0;
+    inst.cities[inst.city].comms = 100;
+    inst.cities[inst.city].pop = 1e9;
+
+    /** Disable random event */
+    inst.randomEventCounter = Infinity;
+  }
+
+  function allCitiesHighChaos() {
+    for (const city of Object.values(inst.cities)) {
+      city.chaos = 1e12;
+    }
+  }
+
+  function complete(id: ActionIdentifier) {
+    start(id), finish();
+  }
+
+  function start(id: ActionIdentifier) {
+    const action = inst.getActionObject(id);
+    if ("count" in action) action.count = 1;
+    if (action.type === BladeburnerActionType.Operation) action.autoLevel = true;
+    if (id.type === "Black Operations") inst.numBlackOpsComplete = (<BlackOperation>action).n;
+    inst.startAction(id);
+  }
+
+  function finish() {
+    inst.processAction(ENOUGH_TIME_TO_FINISH_ACTION);
+    inst.calculateMaxStamina();
+  }
+
+  function* nonGeneralActions() {
+    yield* contracts();
+    yield* operations();
+    yield* Object.values(BlackOperations);
+  }
+
+  function* contracts() {
+    yield* Object.values(instanceUsedForTestGeneration.contracts);
+  }
+
+  function* operations() {
+    yield* Object.values(instanceUsedForTestGeneration.operations);
+  }
+
+  function* actionId(actions: Iterable<Action>) {
+    for (const action of actions) yield { id: action.id };
+  }
+
+  function* actionIdWithIndividualStat(actions: Iterable<Action>) {
+    for (const action of actions) {
+      yield* Object.entries(action.weights)
+        .filter(([__, value]) => value > 0)
+        .map(([stat]) => ({ id: action.id, stat } as { id: ActionIdentifier; stat: keyof Skills }));
+    }
+  }
+});
```

Comment:
These are generators. I am using them solely for the purpose of permutating test cases as they simplify nicely to [...contracts()] and can be composed.

Reply:
Generator functions are a very powerful concept, and they certainly have their place where doing something equivalent without using them requires cumbersome code. However, I don't think that place is here.

In particular, rather than taking and returning Iterables, you can take and return Arrays instead. Then these just become normal functions, still composable, and you don't need [...spread()] syntax when you call them. Also, the `contracts()` and `operations()` functions become trivial to the point of "don't define them" at that point.

Alternatively, you could restructure this functionally. In that case, you'd just remove the `for..of` loops and have these all be functors to be applied with `map()`. Your test definitions would then look like `nonGeneralActions().map(actionIdWithIndividualStat)`